### PR TITLE
worker protocol: serialise cgroup stats in `BuildResult`

### PR DESCRIPTION
**Pull request by r-vdp**
# Motivation

By doing so, they get reported when building through the daemon via either `unix://` or `ssh-ng://`.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libutil/serialise.hh [hunk 1]
```diff
@@ -368,6 +370,10 @@ Sink & operator << (Sink & sink, std::string_view s);
 Sink & operator << (Sink & sink, const Strings & s);
 Sink & operator << (Sink & sink, const StringSet & s);
 
+inline Sink & operator << (Sink & sink, const std::optional<std::chrono::microseconds> & duration) {
+    return sink << (duration.has_value() ? duration.value().count() : -1l);
+}
+
```
Comment (Ericson2314) [inline src/libutil/serialise.hh:376]: Please don't add more functionality here, this should go in `worker-proto.cc` for just that one, or `common-proto.cc` to use in both (e.g. in a follow-up PR for that one if we decide to do it you can move it there)

While `-1` is an unlikely duration to occur in practice, I would use an optional tag instead see 57d960dcd174f0910c7fca02ee3afe80cb2b7844 for example.

File: src/libutil/serialise.hh [hunk 2]
```diff
@@ -420,6 +426,13 @@ Source & operator >> (Source & in, bool & b)
     return in;
 }
 
+template<>
+inline Source & operator >> (Source & in, std::optional<std::chrono::microseconds> & duration) {
+    int64_t value = readNum<uint64_t>(in);
+    duration = value != -1l ? std::optional<std::chrono::microseconds>{std::chrono::microseconds(value)} : std::nullopt;
+    return in;
+}
+
```
Comment (Ericson2314) [inline src/libutil/serialise.hh:435]: ditto from above.

File: src/libutil/serialise.hh (hunk 1 above)
Comment (r-vdp) [reply src/libutil/serialise.hh:376]: But that commit removes the tags in favour of using an empty string for the null option, right?

You meant that you'd prefer that we use an empty string as well, or that we re-introduce tags to distinguish between the two variants?

I used strings initially, but then went for integers to have a more efficient binary representation.

Comment (r-vdp) [reply src/libutil/serialise.hh:376]: Oh, I saw how optional is dealt with in other serialisers in worker-protocol.cc, so I'll do it similarly. Clearly I'm still getting to know the codebase. Thanks for the pointers!

**Comment by Ericson2314** [comment]
OK looking good! I think we just need a release note

File: src/libstore/worker-protocol.cc [hunk 3]
```diff
@@ -47,6 +48,33 @@ void WorkerProto::Serialise<std::optional<TrustedFlag>>::write(const StoreDirCon
 }
 
 
+template<>
+std::optional<std::chrono::microseconds> WorkerProto::Serialise<std::optional<std::chrono::microseconds>>::read(const StoreDirConfig & store, WorkerProto::ReadConn conn)
```
Comment (Ericson2314) [inline src/libstore/worker-protocol.cc:52]: Oh you should declare this in `worker-protocol.hh` too

File: src/libstore/worker-protocol.cc [hunk 4]
```diff
@@ -47,6 +48,33 @@ void WorkerProto::Serialise<std::optional<TrustedFlag>>::write(const StoreDirCon
...
+            return std::nullopt;
+        case 1:
+            return std::optional<std::chrono::microseconds>{std::chrono::microseconds(readNum<int64_t>(conn.from))};
+        default:
+            throw Error("Invalid optional tag from remote");
+    }
+}
+
+template<>
+void WorkerProto::Serialise<std::optional<std::chrono::microseconds>>::write(const StoreDirConfig & store, WorkerProto::WriteConn conn, const std::optional<std::chrono::microseconds> & optDuration)
```
Comment (Ericson2314) [inline src/libstore/worker-protocol.cc:66]: ditto

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**